src/vec.rs

```diff
@@ -984,12 +984,15 @@ impl<T: Clone, const N: usize> From<[T; N]> for EcoVec<T> {
 }
 
 impl<T: Clone> From<Vec<T>> for EcoVec<T> {
-    /// This needs to allocate to change the layout.
-    fn from(other: Vec<T>) -> Self {
-        let mut vec = Self::new();
+    /// Allocates a new EcoVec, and moves other's items into it.
+    fn from(mut other: Vec<T>) -> Self {
+        let mut vec = Self::with_capacity(other.len());
         unsafe {
-            // Safety: Vec's IntoIter implements `TrustedLen`.
-            vec.extend_from_trusted(other);
+            ptr::copy_nonoverlapping(other.as_mut_ptr(), vec.data_mut(), other.len());
```

Comment:
I think doing the set_len first would be slightly safer as it risks a memory leak rather a double drop in case the code in between panics (I don't think that can happen in practice, but still). I.e. like this:
```rs
let len = other.len();
other.set_len(0);
ptr::copy_nonoverlapping(other.as_mut_ptr(), vec.data_mut(), len);
vec.len = len;
```

Could you also add safety comments to the two unsafe function calls explaining why all their individual preconditions are satisfied, similarly to how it's done across this file? Thanks!